### PR TITLE
[mlir][IR] Add listener notifications for pattern begin/end

### PR DESCRIPTION
**Pull request by matthias-springer**
This commit adds two new notifications to `RewriterBase::Listener`:
* `notifyPatternBegin`: Called when a pattern application begins during a greedy pattern rewrite or dialect conversion.
* `notifyPatternEnd`: Called when a pattern application finishes during a greedy pattern rewrite or dialect conversion.

The listener infrastructure already provides a `notifyMatchFailure` callback that notifies about the reason for a pattern match failure. The two new notifications provide additional information about pattern applications.

This change is in preparation of improving the handle update mechanism in the `apply_conversion_patterns` transform op.

### CODE REVIEW

File: mlir/include/mlir/Rewrite/PatternApplicator.h [hunk 1]
```diff
@@ -68,9 +68,9 @@ class PatternApplicator {
   ///            invalidate the match and try another pattern.
   LogicalResult
   matchAndRewrite(Operation *op, PatternRewriter &rewriter,
-                  function_ref<bool(const Pattern &)> canApply = {},
-                  function_ref<void(const Pattern &)> onFailure = {},
-                  function_ref<LogicalResult(const Pattern &)> onSuccess = {});
+                  std::function<bool(const Pattern &)> canApply = {},
+                  std::function<void(const Pattern &)> onFailure = {},
+                  std::function<LogicalResult(const Pattern &)> onSuccess = {});
```
Comment (joker-eph) [inline mlir/include/mlir/Rewrite/PatternApplicator.h:73]: Why this change?

File: mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp [hunk 2]
```diff
@@ -562,30 +562,39 @@ bool GreedyPatternRewriteDriver::processWorklist() {
...
-    auto onSuccess = [&](const Pattern &pattern) {
-      LLVM_DEBUG(logResult("success", "pattern applied successfully"));
-      return success();
-    };
-#else
-    function_ref<bool(const Pattern &)> canApply = {};
-    function_ref<void(const Pattern &)> onFailure = {};
-    function_ref<LogicalResult(const Pattern &)> onSuccess = {};
-#endif
+    debugBuild = true;
```
Comment (joker-eph) [inline mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:570]: Why changing the structure of the code with this variable?

Comment (joker-eph) [reply mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:570]: Oh never mind I see!

File: mlir/include/mlir/Rewrite/PatternApplicator.h (hunk 1 above)
Comment (matthias-springer) [reply mlir/include/mlir/Rewrite/PatternApplicator.h:73]: There is a `stack-use-after-scope` (reported by ASAN, also crashes in opt mode) with `function_ref`. The callback in the greedy pattern rewriter is defined inside of an `if` check:
```c++
    function_ref<bool(const Pattern &)> canApply = {};
    function_ref<void(const Pattern &)> onFailure = {};
    function_ref<LogicalResult(const Pattern &)> onSuccess = {};
    bool debugBuild = false;
#ifdef NDEBUG
    debugBuild = true;
#endif // NDEBUG
    if (debugBuild || config.listener) {
      canApply = [&](const Pattern &pattern) {
        if (this->config.listener) { /* ... */ }
        // ...
      }
      // ...
    }

    // `canApply` points to a lambda that is out of scope.
    LogicalResult matchResult =
        matcher.matchAndRewrite(op, *this, canApply, onFailure, onSuccess);
```

`function_ref` is a "non-owning function wrapper", but the lambda captures `this`.

Changing to `std::function` is one way to fix it. I could also just always pass a lambda. That would actually be my preferred solution, but there is a slight overhead when running in opt mode and without listener because the callback would always be called (even if it does not do anything):
```c++
    LogicalResult matchResult = matcher.matchAndRewrite(
        op, *this,
        /*canApply=*/[&](const Pattern &pattern) {
          if (this->listener) { /* ... */ }
          // ...
        },
        /*onFailure=*/[&](const Pattern &pattern) { /* ... */},
        /*onSuccess=*/[&](const Pattern &pattern) { /* ... */});
```

What do you think?

Comment (matthias-springer) [reply mlir/include/mlir/Rewrite/PatternApplicator.h:73]: (The existing code seemed to care about performance here; There were different `canApply` etc. depending on `NDEBUG`.)

Comment (joker-eph) [reply mlir/include/mlir/Rewrite/PatternApplicator.h:73]: That can explain why you changed it at the call-site, but I'm puzzled about this function: it does not capture the callback as far as I can tell.

Comment (joker-eph) [reply mlir/include/mlir/Rewrite/PatternApplicator.h:73]: > What are you referring to with this function?

Where this comment thread is anchored: `matchAndRewrite`

> The problem here is really just caused by the fact that the canApply = assignment is inside of a nested scope. And the lambda object is dead by the time matcher.matchAndRewrite is called. I.e., the canApply function_ref points to an already free'd lambda. At least that's my understanding.

Yes, but that's a problem for the call-site, I don't quite see where you make the connection to the signature of `matchAndRewrite`?

Comment (matthias-springer) [reply mlir/include/mlir/Rewrite/PatternApplicator.h]: That's a good way to think about it.

File: mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp [hunk 3]
```diff
@@ -572,20 +571,33 @@ bool GreedyPatternRewriteDriver::processWorklist() {
...
+
+#ifdef NDEBUG
+    // Optimization: PatternApplicator callbacks are not needed when running in
+    // optimized mode and without a listener.
+    if (!config.listener) {
+      canApply = nullptr;
+      onFailure = nullptr;
+      onSuccess = nullptr;
+    }
+#endif // NDEBUG
```
Comment (joker-eph) [inline mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:600]: Note: I didn't suggest changing this, what you had here was reasonable!

**Review by joker-eph — APPROVED** [review]
